File: src/__tests__/recipe-response.test.ts
import { Recipe, RecipeResponse, PagedRecipeResponse, AuthorDto } from '../types/recipe';

describe('RecipeResponse DTO contracts', () => {
  it('should instantiate a valid RecipeResponse matching the backend response contract', () => {
    const response: RecipeResponse = {
      id: 'recipe-123',
      userId: 'user-456',
      recipeName: 'Spaghetti Carbonara',
      title: 'Spaghetti Carbonara',
      description: 'Classic Italian pasta',
      ingredients: ['Spaghetti', 'Eggs', 'Pecorino', 'Guanciale'],
      instructions: ['Boil pasta', 'Fry guanciale', 'Mix with eggs and cheese'],
      servings: 4,
      source: 'manual',
      isPublic: true,
      authorDisplayName: 'Chef Mario',
      authorAvatarUrl: 'https://example.com/mario.jpg',
      author: {
        uid: 'user-456',
        displayName: 'Chef Mario',
        avatarUrl: 'https://example.com/mario.jpg'
      },
      likeCount: 42,
      isLikedByCurrentUser: true,
      isSavedByCurrentUser: false,
      averageRating: 4.8,
      ratingCount: 15
    };

    expect(response.id).toBe('recipe-123');
    expect(response.recipeName).toBe('Spaghetti Carbonara');
    expect(response.title).toBe('Spaghetti Carbonara');
    expect(response.authorDisplayName).toBe('Chef Mario');
    expect(response.author?.displayName).toBe('Chef Mario');
    expect(response.likeCount).toBe(42);
    expect(response.isLikedByCurrentUser).toBe(true);
  });

  it('should instantiate a valid PagedRecipeResponse', () => {
    const paged: PagedRecipeResponse = {
      recipes: [
        {
          id: 'r1',
          recipeName: 'Pizza Margherita',
          ingredients: ['Dough', 'Tomato', 'Mozzarella'],
          instructions: ['Bake at 450F'],
          servings: 2,
          source: 'manual',
          authorDisplayName: 'Luigi'
        }
      ],
      size: 1,
      totalCount: 10,
      nextPageToken: 'cursor-token-abc'
    };

    expect(paged.recipes).toHaveLength(1);
    expect(paged.size).toBe(1);
    expect(paged.totalCount).toBe(10);
    expect(paged.nextPageToken).toBe('cursor-token-abc');
  });
});

File: src/test/java/com/recipe/shared/model/RecipeResponseTest.java
package com.recipe.shared.model;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.Arrays;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Parity and serialization tests for RecipeResponse, AuthorDto, and PagedRecipeResponse.
 */
class RecipeResponseTest {

    private final ObjectMapper objectMapper = new ObjectMapper()
            .registerModule(new JavaTimeModule());

    @Test
    void testRecipeResponseSerializationAndDeserialization() throws Exception {
        AuthorDto author = AuthorDto.builder()
                .uid("author-123")
                .displayName("Chef Gordon")
                .avatarUrl("https://example.com/avatar.jpg")
                .build();

        RecipeResponse response = RecipeResponse.builder()
                .id("recipe-1")
                .userId("author-123")
                .recipeName("Pasta Primavera")
                .title("Pasta Primavera")
                .description("Fresh garden pasta")
                .ingredients(Arrays.asList("pasta", "vegetables", "olive oil"))
                .instructions(Collections.singletonList("Boil pasta and toss with vegetables"))
                .servings(2)
                .source("manual")
                .createdAt(Instant.parse("2026-01-01T12:00:00Z"))
                .publicRecipe(true)
                .authorDisplayName("Chef Gordon")
                .authorAvatarUrl("https://example.com/avatar.jpg")
                .author(author)
                .likeCount(42)
                .likedByCurrentUser(true)
                .savedByCurrentUser(false)
                .build();

        String json = objectMapper.writeValueAsString(response);
        assertNotNull(json);
        assertTrue(json.contains("\"isPublic\":true"));
        assertTrue(json.contains("\"isLikedByCurrentUser\":true"));
        assertTrue(json.contains("\"isSavedByCurrentUser\":false"));
        assertTrue(json.contains("\"authorDisplayName\":\"Chef Gordon\""));

        RecipeResponse deserialized = objectMapper.readValue(json, RecipeResponse.class);
        assertEquals("recipe-1", deserialized.getId());
        assertEquals("Pasta Primavera", deserialized.getRecipeName());
        assertEquals("Pasta Primavera", deserialized.getTitle());
        assertEquals("Pasta Primavera", deserialized.getEffectiveTitle());
        assertEquals("Chef Gordon", deserialized.getAuthorDisplayName());
        assertNotNull(deserialized.getAuthor());
        assertEquals("author-123", deserialized.getAuthor().getUid());
        assertEquals(42, deserialized.getLikeCount());
        assertTrue(deserialized.getLikedByCurrentUser());
        assertFalse(deserialized.getSavedByCurrentUser());
    }

    @Test
    void testEffectiveTitleFallback() {
        RecipeResponse titleOnly = RecipeResponse.builder().title("Title Only").build();
        assertEquals("Title Only", titleOnly.getEffectiveTitle());

        RecipeResponse recipeNameOnly = RecipeResponse.builder().recipeName("Recipe Name Only").build();
        assertEquals("Recipe Name Only", recipeNameOnly.getEffectiveTitle());

        RecipeResponse both = RecipeResponse.builder().title("Title").recipeName("Recipe Name").build();
        assertEquals("Title", both.getEffectiveTitle());
    }

    @Test
    void testPagedRecipeResponse() throws Exception {
        RecipeResponse item = RecipeResponse.builder()
                .id("rec-1")
                .recipeName("Dish 1")
                .build();

        PagedRecipeResponse paged = PagedRecipeResponse.builder()
                .recipes(Collections.singletonList(item))
                .size(1)
                .totalCount(100)
                .nextPageToken("token-abc")
                .build();

        String json = objectMapper.writeValueAsString(paged);
        PagedRecipeResponse deserialized = objectMapper.readValue(json, PagedRecipeResponse.class);

        assertEquals(1, deserialized.getSize());
        assertEquals(100, deserialized.getTotalCount());
        assertEquals("token-abc", deserialized.getNextPageToken());
        assertEquals(1, deserialized.getRecipes().size());
        assertEquals("Dish 1", deserialized.getRecipes().get(0).getRecipeName());
    }
}
